### PR TITLE
Reduce number of `prim::profile`

### PR DESCRIPTION
**Pull request by Krovatkin**
Fixes #{issue number}

### CODE REVIEW

**Review by ZolotukhinM — COMMENTED** [review]
Could you please add a brief description (in the commit message) of how and what this fixes?

File: torch/csrc/jit/runtime/profiling_record.cpp [hunk 1]
```diff
@@ -194,10 +195,18 @@ bool needsProfiledOutput(Node* n) {
...
   for (auto it = block->nodes().begin(); it != block->nodes().end(); ++it) {
     auto n = *it;
-    for (auto i : n->inputs()) {
+    for (size_t offset = 0; offset < n->inputs().size(); offset++) {
+      auto i = n->input(offset);
+      // n can have multiple uses, if we already
+      // processed `n` we also inserted `prim::profile`s
+      // let's check for this case otherwise we will
+      // be `prim::profile` `prim::profile`
+      if (i->node()->kind() == prim::profile) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/runtime/profiling_record.cpp:204]: Does it ever fire? I suspect that it shouldn't once we changed the way we replace inputs (previously we were replacing all inputs of a given value, now we're only replacing one specific input with a given index).

**Review by ZolotukhinM — APPROVED** [review]
Looks good, but please add some description of what the problem was and also please double check if a check `i->node()->kind() == prim::profile` in needed (see inline).

**Review by eellison — COMMENTED** [review]
Don't we want different profiles for different subblocks ?
```
def foo(x):
    if x.size() == 2:
          print(x)
    else:
          print(x)
```
In this case, the usage of `print(x)` will yield a different profile than the parent profile, so it's actually providing new information.

I think prim::profiles that are on the same block are redundant, but i'm not sure about other usages ( and even then, do we need to worry about mutating ops ) ?

**Comment by Krovatkin** [comment]
> Don't we want different profiles for different subblocks ?

This PR isn't changing whether different uses of the same definition will get the same prim::profile or different ones. We are still inserting `prim::profile` on every use, so we should be getting different profiles for different subblocks.

> I think prim::profiles that are on the same block are redundant, but i'm not sure about other usages ( and even then, do we need to worry about mutating ops ) 

I don't think profiler should be responsible for figuring out which `prim::profile` redundant and which one aren't. It should profile every use which implicitly handles the mutating ops case to provide the most accurate information.

**Comment by eellison** [comment]
It might be nice to just add a special case to dead code elimination instead, so that graphs which are cleaned up and have adjacent prim::profiles after the initial insertion also get cleaned up

File: torch/csrc/jit/runtime/profiling_record.cpp (hunk 1 above)
Comment (Krovatkin) [reply torch/csrc/jit/runtime/profiling_record.cpp:204]: yeah, giving this some more thought, it **shouldn't** fire. I also double checked by adding an assertion in that if and it never fired (modulo windows failures) which don't seem to be related to my changes.

**Comment by Krovatkin** [comment]
> It might be nice to just add a special case to dead code elimination instead, so that graphs which are cleaned up and have adjacent prim::profiles after the initial insertion also get cleaned up

this is a great idea! why DCE though and not a peephole? They aren't dead per se, just redundant.

**Pull request closed without merging**